### ISSUE TITLE
add TreeNode.from_linkage/to_linkage methods

### ISSUE THREAD

**Issue by gregcaporaso** [biocore/scikit-bio#346]
The [scipy linkage object](http://docs.scipy.org/doc/scipy-0.13.0/reference/generated/scipy.cluster.hierarchy.linkage.html) is returned by the scipy hierarchical clustering methods, so being able to load or generate these from TreeNode's provides the glue between scipy's hierarchical clustering and skbio's TreeNode. 

@ebolyen is working on this. @wasade, would you be available to review?

**Comment by mortonjt** [comment]
Adding the `to_linkage` method would be extremely helpful.  For example, it would be really nice to pass a linkage matrix into [`seaborn.clustermap`](https://stanford.edu/~mwaskom/software/seaborn/generated/seaborn.clustermap.html) to align a tree with a heatmap.

**Comment by jairideout** [comment]
Agree, is anyone available to work on `to_linkage`?

**Comment by mortonjt** [comment]
I'll need this for a paper anyways, so I'll probably have something ready within a month.

**Comment by jairideout** [comment]
Awesome! Is it okay if I assign this issue to you?

**Comment by mortonjt** [comment]
Sure thing.

**Comment by gregcaporaso** [comment]
This will be really great to get in there.

**Comment by mortonjt** [comment]
Hi @gregcaporaso @jairideout .  I've been looking at this more carefully.

It turns out that this is a really hard problem.
1. The linkage format assumes balanced distances (i.e. the distance between children and their parents are equal), it is not possible to actually encode the branch lengths into the linkage matrix
2. The encoding isn't very clear - but if I have to guess based on the source code, it looks like there is a reverse post-traversal going on (right node gets traversed first instead of the left node).  

Sorry to be the bear of bad news - but it looks like we would lose the branch length information if we cast to this object.  But if we really wanted this, it is probably possible to do this cast by setting all of the distances to 1 or something.  Even with this scheme, it is tricky to figure out how to get the right matrix, given the traversal issue.

**Comment by ebolyen** [comment]
@mortonjt, branch lengths can be encoded, and siblings do not have to have identical distances to their parents (except for tips), however the tree *does* have to be an ultrametric binary tree (strictly bifurcating with every tip being equidistant to the root). 

The linkage isn't exactly a reversed depth-first search, it actually is ordered by shortest subtree (you should see distances increase as you traverse the linkage matrix as the distance recorded represents the tip to root so far).

I think we should have some way of validating that a given tree is bifurcating and ultrametric, and raise an error otherwise.

**Comment by mortonjt** [comment]
@ebolyen you are right - this method will only work properly on ultrametric bifurcating trees.  I didn't know about the shortest subtree ordering.  Thanks!

**Comment by mortonjt** [comment]
This may of interest in the future -- particularly now that seaborn remains to be the best plotting program for dendrogram heatmaps.
